extension/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+	dispatch::{GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::OriginTrait,
+};
+use frame_system::RawOrigin;
+use pallet_contracts::chain_extension::{
+	BufInBufOutState, ChainExtension, Environment, Ext, InitState, RetVal,
+};
+use sp_core::crypto::UncheckedFrom;
+use sp_runtime::{traits::Dispatchable, DispatchError};
+use sp_std::vec::Vec;
+
+type ContractSchedule<T> = <T as pallet_contracts::Config>::Schedule;
+
+/// Handles the query from the chain extension environment for state reads.
+pub trait ReadState {
+	type StateQuery: Decode;
+
+	/// Allowed state queries from the API.
+	fn contains(c: &Self::StateQuery) -> bool;
+
+	/// Reads state using the provided query, returning the result as a byte vector.
+	fn read(read: Self::StateQuery) -> Vec<u8>;
+
+	/// Decodes parameters into state query.
+	fn decode(params: &mut &[u8]) -> Result<Self::StateQuery, DispatchError> {
+		decode_checked(params)
+	}
+}
+
+/// Handles the query from the chain extension environment for dispatch calls.
+pub trait CallFilter {
+	type Call: Decode;
+
+	/// Allowed runtime calls from the API.
+	fn contains(t: &Self::Call) -> bool;
+}
+
+#[derive(Default)]
+pub struct ApiExtension<I>(PhantomData<I>);
+
+impl<T, I> ChainExtension<T> for ApiExtension<I>
+where
+	T: pallet_contracts::Config
+		+ frame_system::Config<
+			RuntimeCall: GetDispatchInfo + Dispatchable<PostInfo = PostDispatchInfo>,
+		>,
+	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+	// Bound the type by the two traits which need to be implemented by the runtime.
+	I: ReadState + CallFilter<Call = <T as frame_system::Config>::RuntimeCall> + 'static,
+{
+	fn call<E: Ext<T = T>>(
+		&mut self,
+		env: Environment<E, InitState>,
+	) -> Result<RetVal, DispatchError> {
+		log::debug!(target:LOG_TARGET, " extension called ");
+		let mut env = env.buf_in_buf_out();
+		// Charge weight for making a call from a contract to the runtime.
+		// `debug_message` weight is a good approximation of the additional overhead of going
+		// from contract layer to substrate layer.
+		// reference: https://github.com/paritytech/ink-examples/blob/b8d2caa52cf4691e0ddd7c919e4462311deb5ad0/psp22-extension/runtime/psp22-extension-example.rs#L236
+		let contract_host_weight = ContractSchedule::<T>::get().host_fn_weights;
+		env.charge_weight(contract_host_weight.debug_message)?;
+
+		let (version, function_id, pallet_index, call_index) = extract_env(&env);
+
+		let result = match FuncId::try_from(function_id) {
+			// Read encoded parameters from buffer and calculate weight for reading `len` bytes`.
+			Ok(function_id) => {
+				// reference: https://github.com/paritytech/polkadot-sdk/blob/117a9433dac88d5ac00c058c9b39c511d47749d2/substrate/frame/contracts/src/wasm/runtime.rs#L267
+				let len = env.in_len();
+				env.charge_weight(contract_host_weight.return_per_byte.saturating_mul(len.into()))?;
+				let params = env.read(len)?;
+				log::debug!(target: LOG_TARGET, "Read input successfully");
```

Comment:
Do we still need this? Seems redundant. Perhaps it should be augmented with some state to be a bit more useful.